### ISSUE TITLE
`-l,--location` option not working on all msi packages

### ISSUE THREAD

**Issue by Cube707** [microsoft/winget-cli#1857]
### Brief description of your issue

This issue is a followup [on my PR for a package mainfest](https://github.com/microsoft/winget-pkgs/pull/40984).

There seem to be a lot of inconsistencyes on how to set the install location path for msi packages. The code currently sets the `TARGETDIR` option.

https://github.com/microsoft/winget-cli/blob/b79e4087821c993f83f7c4731bbde7aa25612c73/src/AppInstallerCommonCore/Manifest/ManifestCommon.cpp#L461

This is only one of two available options ([see here](https://stackoverflow.com/a/8841965)) and some packages seem to require the alternative `INSTALLDIR` option.

I have tested the packages `Oracle.JDK.17`, `StrawberryPerl.StrawberryPerl` and `ChristianHohnstadt.xca`, which all currently don't work with the `-l` switch.

In contrast to that, the `Borvid.HttpMasterExpress` package works absoloutly fine as is.

One possible soloution would be to allways pass both of these options, which worked in my short testing.

### Steps to reproduce

Use one of the following commands:

`winget install Oracle.JDK.17 -l C:\ProgStuff\jdk`
`winget install Oracle.JDK.17 -l "C:\ProgStuff\jdk"`
`winget install Oracle.JDK.17 --override "/passive TARGETDIR=C:\ProgStuff\jdk"`

The jdk is now installed into `C:\Program Files\Java\jdk17` and *NOT* into the given location.

Only `winget install Oracle.JDK.17 --override "/passive INSTALLDIR=C:\ProgStuff\jdk"` works as expected.

### Expected behavior

The package should be installed into the given Path.

### Actual behavior

The given directory is ignored and the package is installed into the default directory.

### Environment

```shell
[winget --info]
Windows Package Manager v1.1.13405
Copyright (c) Microsoft Corporation. Alle Rechte vorbehalten.

Windows: Windows.Desktop v10.0.19043.1415
Paket: Microsoft.DesktopAppInstaller v1.16.13405.0

Protokolle: %LOCALAPPDATA%\Packages\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe\LocalState\DiagOutputDir
```

**Comment by OfficialEsco** [comment]
Sooo i learned something today in https://github.com/microsoft/winget-pkgs/pull/41913#discussion_r790846573
I would bet ANY installer that does not respect `-l` is actually `wix` instead of `msi` 🤦‍♂️🤦‍♂️🤦‍♂️
I don't know my way around this repo, but this tends to be where my searches lead me https://github.com/microsoft/winget-cli/blob/b79e4087821c993f83f7c4731bbde7aa25612c73/src/AppInstallerCommonCore/Manifest/ManifestCommon.cpp#L451-L481
And from my understand winget-cli uses the same switches for burn, wix and msi, if someone could double check the InstallerSwitches for burn and wix we over at the winget-pkgs could fix the InstallerTypes

cc @ItzLevvie @jedieaston @Trenly @denelon

**Comment by Trenly** [comment]
Apparently they should be the same?
https://stackoverflow.com/a/44347702/9998654

**Comment by jedieaston** [comment]
It looks like WiX doesn't have a standard argument for install location, but by convention people use INSTALLLOCATION? https://stackoverflow.com/questions/24266720/how-to-define-installation-folder-from-command-line-parameter-in-wix-installer

**Comment by dorssel** [comment]
Just my 2 cents.

WiX has always used something different than TARGETDIR. The reason is that WiX actually does it "wrong" (or at least: their advertised examples).

The MSI specs are very clear: TARGETDIR is to be used if you want the non-standard installation folder. See: https://docs.microsoft.com/en-us/windows/win32/msi/targetdir (first paragraph).

To accomplish this, a MSI *should* set the TARGETDIR property if and only if it is not set by the user. Something like (pseudo code):
```
if TARGETDIR is not set:
   set TARGETDIR = %ProgramFiles%\Vendor\Product
```
This is clear from https://docs.microsoft.com/en-us/windows/win32/msi/using-the-directory-table.

This can be easily accomplished in WiX as well, by a `<SetProperty Id="TARGETDIR" Value="<XXX>" Before="CostInitialize">TARGETDIR == ""</SetProperty>`. However, almost no one actually does this. Most WiX users go for the standard WiX MinimalUI or WiX AdvancedUI templates. Those are really old (10+ years) and not well maintained. Besides, most online examples are equally old and this keeps the "myth alive".

Most WiX examples use:
```
<Directory Id="TARGETDIR" Name="SourceDir">
    <Directory Id="ProgramFiles64Folder">
        <Directory Id="Vendor" />
            <Directory Id="APPLICATIONFOLDER" Name="YYY" />
        </Directory>
    </Directory>
</Directory>
```
The problem is that even if you override TARGETDIR, the next "child directory" (note: that is not the same as a subdirectory) immediately get set to %ProgramFiles64Folder%, as that is a predefined property. As a result, the TARGETDIR gets completely ignored and everything ends up below %ProgramFiles64Folder% again. _**Unless**_ you override the very last directory in this list. In the example above that would be APPLICATIONFOLDER. That is what current WiX examples use: https://wixtoolset.org//documentation/manual/v3/wixui/dialog_reference/wixui_advanced.html. But in the past, it has been INSTALLLOCATION. And in fact, it could be anything. Even worse: if the Id of the final destination directory is not spelled in ALL-CAPPS, then you cannot even override it at all.

To be clear, the above example *should* read:
```
<SetProperty Id="TARGETDIR" Value="<enter default here>" Before="CostInitialize">NOT TARGETDIR</SetProperty>
<Directory Id="TARGETDIR" Name="SourceDir">
```
but (again): nobody actually does this.

**Comment by BinaryWizard904** [comment]
How the `-l` and `--location` options work is defined by the `InstallLocation` field of the [Installer Manifest](https://github.com/microsoft/winget-pkgs/blob/master/doc/manifest/schema/1.1.0/installer.md).  As an example of how to get the `-l` and `--location` options working with WiX installers, check out the [PuTTY manifest](https://github.com/microsoft/winget-pkgs/blob/master/manifests/p/PuTTY/PuTTY/0.77.0.0/PuTTY.PuTTY.installer.yaml).  Specifically lines 14-15:
```
InstallerSwitches:
  InstallLocation: INSTALLDIR="<INSTALLPATH>"
```

In other words, this is not an issue with the client, but rather an issue with incomplete manifest files.

**Comment by ItzLevvie** [comment]
`InstallLocation` is actually defined automatically depending on the `InstallerType` in the manifest: https://github.com/microsoft/winget-cli/blob/5d63c3870e195c9b0ee1618902f6fa7aa43165d1/src/AppInstallerCommonCore/Manifest/ManifestCommon.cpp#L477-L509

^ This is the reason why all MSI / WIX manifests in the winget-pkgs repository don't have it. It's also the same with `InstallerSwitches`.

It does look like some installers work a bit differently—probably because of the differences between WIX and MSI which is why Python uses `INSTALLDIR` instead of `TARGETDIR`.

**Comment by dinhani** [comment]
Some MSI files like Zeal are using `INSTALL_ROOT` to define the install location. Setting the value with `--custom INSTALL_ROOT=path/to/dir` works.

I am not sure if this is a standard property, but it seems there are more applications using it.
